### PR TITLE
First commit to updated CI

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
# Description

This fixes the CI metadata services after using Blake's cleaner docker containers. It also adds the repo services and the Globus instance as part of the CI setup. At the end of the CI end-to-end tests have been also tested.

Diagram to CI pipeline is here.

https://confluence.ccs.ornl.gov/display/DATAFED/FLOW+-+CI

### CODE REVIEW

**Review by nedvedba — APPROVED** [review]

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -550,131 +456,633 @@ end-to-end-core-setup:
...
-      -DBUILD_DOCS=OFF
-      -DBUILD_PYTHON_CLIENT=ON
-      -DBUILD_TESTS=ON
-      -DENABLE_END_TO_END_TESTS=ON
-      -DINSTALL_FOXX=OFF
-    - cmake --build build
-    - cmake --build build --target pydatafed
-    - cmake --build build --target test
 
+#################################################################################
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply .gitlab-ci.yml:518]: Yes, that is correct.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -3,16 +3,21 @@ stages:
...
+# WARNING
+#
+# Do not put if else statements without if and else, gitlab yaml does not like
+# this kind of scripting
+#
+# [ -f file_path ] && ok do this....
 
 variables:
-  HARBOR_USER: 'robot$$datafed+harbor_datafed_gitlab_ci_registry'
+  HARBOR_USER: 'robot$$datafed+harbor_datafed_gitlab_ci_registry_2'
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply .gitlab-ci.yml:20]: The initial client expired and the same name could not be used.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -236,19 +158,19 @@ provision-client:
...
-        - scripts/install_client_dependencies.sh
-        - scripts/dependency_install_functions.sh
-        - scripts/dependency_versions.sh
- 
+      #  rules:
+      #    - changes:
+      #        - scripts/generate_datafed.sh
+      #        - scripts/install_client_dependencies.sh
+      #        - scripts/dependency_install_functions.sh
+      #        - scripts/dependency_versions.sh
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply .gitlab-ci.yml:166]: Good catch.

File: tests/end-to-end/test_api_context.py [hunk 4]
```diff
@@ -52,6 +58,7 @@ def setUp(self):
             assert count < 3
 
     def test_context(self):
+        print("Running getContext()")
```
Comment (nedvedba) [inline tests/end-to-end/test_api_context.py:61]: Debug print

File: tests/end-to-end/test_api_user_login.py [hunk 5]
```diff
@@ -33,15 +33,18 @@ def test_login_with_password(self):
...
         df_api = API(opts)
-
+        print("API created")
         username = "datafed99"
         password = os.environ.get("DATAFED_USER99_PASSWORD")
 
         count = 0
         while True:
             try:
+                print(f"username: {username}, password: {password}")
```
Comment (nedvedba) [inline tests/end-to-end/test_api_user_login.py:47]: I assume that this is just because it is a test, but it is printing a password.

**Review by nedvedba — APPROVED** [review]
Overall, I think this is pretty close to ready, it may be beneficial to make some sort of a diagram of how the CI runs/test things for clarity.

File: tests/end-to-end/test_api_context.py (hunk 4 above)
Comment (JoshuaSBrown) [reply tests/end-to-end/test_api_context.py:61]: Good catch.

File: tests/end-to-end/test_api_user_login.py (hunk 5 above)
Comment (JoshuaSBrown) [reply tests/end-to-end/test_api_user_login.py:47]: Goot catch, this should be removed though.

File: scripts/dependency_install_functions.sh
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply scripts/dependency_install_functions.sh:1]: Good catch.

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -286,147 +206,149 @@ build-core:
...
-      when: on_success
   script:
     - BRANCH_LOWER=$(echo "$CI_COMMIT_REF_NAME" | tr '[:upper:]' '[:lower:]')
     - echo "$BRANCH_LOWER"
-    - env
     - ./scripts/generate_datafed.sh
-    - cat ./config/datafed.sh
     - docker login "${REGISTRY}" -u "${HARBOR_USER}" -p "${HARBOR_DATAFED_GITLAB_CI_REGISTRY}"
-    - docker build --build-arg BASE_CORE_IMAGE="${REGISTRY}/${IMAGE_TAG}-base-${BRANCH_LOWER}:latest" -f core/docker/Dockerfile.core.ubuntu -t "${REGISTRY}/${IMAGE_TAG}-${BRANCH_LOWER}" .
+    - docker build --build-arg DEPENDENCIES="${REGISTRY}/datafed/dependencies-${BRANCH_LOWER}:latest" --build-arg RUNTIME="${REGISTRY}/datafed/runtime-${BRANCH_LOWER}:latest" -f core/docker/Dockerfile -t "${REGISTRY}/${IMAGE_TAG}-${BRANCH_LOWER}:latest" .
```
Comment (JoshuaSBrown) [inline .gitlab-ci.yml:222]: @nedvedba I want you to point things out like this, there is a lot of repetition between the different docker builds, and also in how the containers are run.  From an operations point of view it makes sense to have those items in a operational type scripts so someone can just pick the script up and run it.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -286,147 +206,149 @@ build-core:
...
+    - ci-datafed-repo
+    - docker
+  rules:
+    - changes:
+        - docker/**/*
+        - scripts/**/*
+        - web/**/*
+        - common/proto/**/*
+        - .gitlab-ci.yml
+      when: on_success
```
Comment (JoshuaSBrown) [inline .gitlab-ci.yml:241]: Another thing you could have pointed out, there is probably room for optimization around the rules. Did you look at these in any detail?

**Comment by JoshuaSBrown** [comment]
@nedvedba did you look at any of the build pipelines and check to see if you understood how the CI works, or have any suggestions for improvements?

**Pull request merged**